lib/train-winrm/socks_proxy_patch.rb

```diff
@@ -3,33 +3,107 @@
 # through a SOCKS proxy by overriding the socket creation method.
-# It also configures `TCPSocket` to use the specified SOCKS proxy server.
+# It also configures `TCPSocket` to use the specified SOCKS proxy server,
+# and validates that the provided proxy is correct, resolvable, and accessible.
+
+require "socksify"
+require "httpclient"
+require "socket" unless defined?(Socket)
 
 class SocksProxyPatch
   # Applies the SOCKS proxy settings to `HTTPClient` and `TCPSocket`.
   #
   # @param socks_proxy [String] The SOCKS proxy address in the format `host:port`.
+  # @param socks_user [String, nil] Optional SOCKS proxy username.
+  # @param socks_password [String, nil] Optional SOCKS proxy password.
   # @example
-  #   SocksProxyPatch.apply("127.0.0.1:1080")
-  def self.apply(socks_proxy, socks_user: nil, socks_password: nil)
-    require "socksify"
-    require "httpclient"
+  #   SocksProxyPatch.apply(socks_proxy: "127.0.0.1:1080", socks_user: "user", socks_password: "pass")
+  def self.apply(socks_proxy:, socks_user: nil, socks_password: nil)
+    new(socks_proxy: socks_proxy, socks_user: socks_user, socks_password: socks_password).apply
+  end
+
+  def initialize(socks_proxy:, socks_user:, socks_password:)
+    @socks_proxy = socks_proxy
+    @socks_user = socks_user
+    @socks_password = socks_password
+  end
+
+  def apply
+    @proxy_host, @proxy_port = parse_and_validate_proxy(@socks_proxy)
+    configure_socks
+    patch_http_client
+  end
 
-    # Extract the host and port from the SOCKS proxy address.
+  private
+
+  attr_reader :proxy_host, :proxy_port, :socks_user, :socks_password
+
+  # Parses the proxy string and validates its format, DNS resolution, and port.
+  def parse_and_validate_proxy(socks_proxy)
     proxy_host, proxy_port = socks_proxy.split(":")
-    TCPSocket.socks_server = proxy_host
-    TCPSocket.socks_port   = proxy_port.to_i
+    proxy_host = proxy_host.to_i
+
+    unless proxy_host && proxy_port
+      raise Train::ClientError, "Invalid SOCKS proxy format: '#{socks_proxy}'. Expected format is 'host:port'."
+    end
+
+    validate_dns_resolution(proxy_host)
+    port = validate_port(proxy_port)
+
+    [proxy_host, port]
+  end
+
+  # Checks if the hostname is resolvable.
+  def validate_dns_resolution(host)
+    Socket.getaddrinfo(host, nil)
+  rescue SocketError => e
+    raise Train::ClientError, "DNS resolution failed for SOCKS proxy host '#{host}': #{e.message}"
+  end
+
+  # Ensures port is a valid integer in range.
+  def validate_port(port_str)
+    port = Integer(port_str)
+    unless port.between?(1, 65_535)
+      raise Train::ClientError, "SOCKS proxy port '#{port}' is out of valid range (1-65535)."
+    end
+
+    port
+  rescue ArgumentError
+    raise Train::ClientError, "Invalid SOCKS proxy port '#{port_str}'. Port must be an integer."
+  end
+
+  # Applies the validated proxy settings to TCPSocket.
+  def configure_socks
     TCPSocket.socks_username = socks_user if socks_user
```

Comment:
If the `socks_user` and `socks_password` are not provided, should this still work without them?